contracts/NFTMarketplace.sol

```diff
@@ -0,0 +1,197 @@
+
+    constructor(address _marketItemAddress) {
+        marketItem = MarketItem(_marketItemAddress);
+        owner = payable(msg.sender);
+    }
+
+    modifier HasTransferApproval(uint256 _tokenId) {
+        require(marketItem.getApproved(_tokenId) == address(this), 'Market is not approved');
+        _;
+    }
+
+    modifier IsForSale(uint256 _tokenId) {
+        require(!nftLedger[_tokenId].forSale, 'Item is already sold');
+        _;
+    }
+
+    function createCollection(string memory _collectionName) external {
+        require(bytes(_collectionName).length != 0, 'Collection cannot be empty');
+
+        _collectionIds.increment();
+        uint256 newCollectionId = _collectionIds.current();
+
+        collectionLedger[newCollectionId] = Collection(_collectionName);
+
+        collectionIds.push(newCollectionId);
+        emit CollectionCreated(newCollectionId, _collectionName);
+    }
+
+    function createMarketItem(uint256 _collectionId, string memory _uri) external {
+        uint256 tokenId = marketItem.safeMint(msg.sender, _uri);
+
+        nftLedger[tokenId] = MarketNft(tokenId, _collectionId, 0, false);
+
+        nftLedgerIds.push(tokenId);
+        emit MarketNftCreated(tokenId, 0, _collectionId, false);
+    }
+
+    function setApproval(address _marketplaceContract, uint256 _tokenId) external {
+        marketItem.approve(_marketplaceContract, _tokenId);
+    }
+
+    function listItem(uint256 _tokenId, uint256 _price) private {
+        // - approval FE marketItem
+        nftLedger[_tokenId].forSale = true;
+        nftLedger[_tokenId].price = _price;
+    }
+
+    function buyItem(uint256 _tokenId)
+        external
+        payable
+        IsForSale(_tokenId)
+        HasTransferApproval(_tokenId)
+    {
+        uint256 price = nftLedger[_tokenId].price;
+
+        require(msg.value >= price, 'Not enough funds sent');
+
+        nftLedger[_tokenId].forSale = true;
+        marketItem.transferFrom(marketItem.ownerOf(_tokenId), msg.sender, _tokenId);
+
+        //  payable(msg.sender).transfer(msg.value);
+        //   owner.transfer(LISTING_FEE);
+
+        //  _allowForPull(marketItem.ownerOf(_tokenId), (msg.value - LISTING_FEE));
+    }
+
+    function _allowForPull(address receiver, uint256 amount) private {
+        credits[receiver] += amount;
+    }
+
+    function withdrawCredits() public {
+        uint256 amount = credits[msg.sender];
+
+        require(amount > 0, 'There are no credits in this recipient address');
+        require(address(this).balance >= amount, 'There are no credits in this contract address');
+
+        credits[msg.sender] = 0;
+
+        payable(msg.sender).transfer(amount);
+    }
+
+    function cancelItem(uint256 _tokenId) public payable {
+        uint256 price = nftLedger[_tokenId].price;
+
+        require(
+            msg.value == price,
+            'Please submit the asking price in order to complete the purchase'
+        );
+
+        nftLedger[_tokenId].forSale = false;
```

Comment:
This whole logic is wrong, when you remove an item from sale, you need to just update the status, and set the price to 0